javacutil/src/main/java/org/checkerframework/javacutil/CollectionUtils.java

```diff
@@ -23,4 +28,203 @@ protected boolean removeEldestEntry(Map.Entry<K, V> entry) {
+    return cloneElements(orig, false);
+  }
+
+  /**
+   * Returns a copy of {@code orig}, where each value of the result is a clone of the corresponding
+   * value of {@code orig}, but the keys are the same objects.
+   *
+   * @param <K> the type of keys of the map
+   * @param <V> the type of values of the map
+   * @param <M> the type of the map
+   * @param orig a map
+   * @return a deep copy of {@code orig}
+   */
+  @SuppressWarnings({"nullness", "signedness"}) // generics problem with clone
+  public static <K, V, M extends Map<K, V>> M cloneValues(M orig) {
+    return cloneElements(orig, false);
+  }
+
+  /**
+   * Returns a copy of {@code orig}, where each key and value in the result is a clone of the
+   * corresponding element of {@code orig}.
+   *
+   * @param <K> the type of keys of the map
+   * @param <V> the type of values of the map
+   * @param <M> the type of the map
+   * @param orig a map
+   * @param cloneKeys if true, clone keys; otherwise, re-use them
+   * @return a deep copy of {@code orig}
+   */
+  @SuppressWarnings({"nullness", "signedness"}) // generics problem with clone
+  private static <K, V, M extends Map<K, V>> M cloneElements(@PolyNull M orig, boolean cloneKeys) {
+    if (orig == null) {
+      return null;
+    }
+    M result = clone(orig);
+    result.clear();
+    for (Map.Entry<K, V> mapEntry : orig.entrySet()) {
+      K oldKey = mapEntry.getKey();
+      K newKey = cloneKeys ? clone(oldKey) : oldKey;
+      result.put(newKey, clone(mapEntry.getValue()));
+    }
+    return result;
+  }
+
+  /**
+   * Returns a copy of {@code orig}, where each element of the result is a deep copy of the
+   * corresponding element of {@code orig}.
+   *
+   * @param <T> the type of elements of the collection
+   * @param <C> the type of the collection
+   * @param orig a collection
+   * @return a deep copy of {@code orig}
+   */
+  @SuppressWarnings({"signedness", "nullness:argument"}) // problem with clone()
+  public static <T extends @Nullable DeepCopyable, C extends @Nullable Collection<T>> @PolyNull C deepCopy(@PolyNull C orig) {
+    if (orig == null) {
+      return null;
+    }
+    C result = clone(orig);
+    result.clear();
+    for (T elt : orig) {
+      @SuppressWarnings("unchecked")
+      T newElt = elt == null ? elt : (T) elt.deepCopy();
+      result.add(newElt);
+    }
+    return result;
+  }
+
+  /**
+   * Returns a copy of {@code orig}, where each key and value in the result is a deep copy of the
+   * corresponding element of {@code orig}.
+   *
+   * @param <K> the type of keys of the map
+   * @param <V> the type of values of the map
+   * @param <M> the type of the map
+   * @param orig a map
+   * @return a deep copy of {@code orig}
+   */
+  @SuppressWarnings({"nullness", "signedness"}) // generics problem with clone
+  public static <
+          K extends @Nullable DeepCopyable, V extends @Nullable DeepCopyable, M extends Map<K, V>>
+      M deepCopy(M orig) {
```

Comment:
Why is `deepCopy` for collections polymorphic, but not this method?

Reply:
> Why is `deepCopy` for collections polymorphic, but not this method?

I think this method is equally polymorphic.  Could you explain what I am missing?

Reply:
Sorry, my comment was referring to qualifier polymorphism (not Java polymorphism). This has been resolved: you added `@PolyNull` annotations.